### PR TITLE
Explain use of pod os field

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
Identify OS authoritatively during pod admission time using `pod.OS` field

### CODE REVIEW

**Comment by sftim** [comment]
/retitle [WIP] Explain use of pod os field

**Comment by katcosgrove** [comment]
Hi from the Comms team! Just a reminder that the Ready to Review deadline for feature blogs is Tuesday, August 16. You will also be assigned a publication date post-release. Is there anything we can do to help you right now?

File: content/en/blog/_posts/2022-07-27-pod-os-field/index.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+As a result, the end user of the kubernetes cluster can set Linux specific security constraints onto Windows pods or vice-versa. 
+These security constraints are passed on to the underlying container runtime from kubelet and since these security constraints 
+are specific to OS, the containers may fail to come up.
+This problem worsens if we have admission plugins or webhooks like `PodSecurity` in our kubernetes clusters which adds the
+security constraints to the pods based on the security profile we choose.
+While the `kubelet` can identify the OS on which it runs and can strip certain security constraints before passing
+onto the container runtime, it'd be better to identify the OS early in the life-cycle of pod and use the 
+identification mechanism consistently across all kubernetes components including `kubelet`. 
+With that in mind, `OS` field has been added to the pod spec in 1.23 release of kubernetes as an 
+alpha feature and it graduated to stable in 1.25.`PodSecurity` admission plugin has been updated to use the `OS` field.
```
Comment (jsturtevant) [inline content/en/blog/_posts/2022-07-27-pod-os-field/index.md:26]:
```suggestion
alpha feature and it graduated to stable in 1.25. In addition, the `PodSecurity` admission plugin has been updated to use the `OS` field. 
```
Or something like that

Mabye a link to the admission plugin?

File: content/en/blog/_posts/2022-07-27-pod-os-field/index.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+---
+
+**Authors:** Ravi Gudimetla (Apple)
+
+This blog describes how to identify pod's operating system authoritatively using the `OS` field in the pod spec and how it 
+helps in securing the kubernetes clusters while ensuring that the workloads can run smoothly.  
+
+## What problems does this solve?
+
+Today, the `kube-apiserver` and `kubelet` while admitting pods have no notion of the OS on which the pod can run.
```
Comment (jsturtevant) [inline content/en/blog/_posts/2022-07-27-pod-os-field/index.md:16]:
```suggestion
Today, at pod admission time the `kube-apiserver` and `kubelet` have no notion of the OS on which the pod can run. 
```

File: content/en/blog/_posts/2022-07-27-pod-os-field/index.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+- SeccompProfile
+when `pod.Spec.OS.Name` is set to `Windows`
+
+
+## How do I use it?
+By setting `pod.Spec.OS.Name=Windows` or `pod.Spec.OS.Name=Linux`
+
+## How can I learn more?
+KEP: https://github.com/kubernetes/enhancements/issues/2802
+API Changes: https://github.com/kubernetes/kubernetes/pull/104693
```
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-pod-os-field/index.md:46]: Once https://kubernetes.io/docs/concepts/workloads/pods/ mentions the new field (see https://github.com/kubernetes/website/pull/35590), please link there.

**Review by sftim — COMMENTED** [review]
Some early feedback.

/hold
as valid a publication date isn't assigned yet

**Comment by katcosgrove** [comment]
Hi there! Your assigned publication date is September 5. Thank you!

**Comment by sftim** [comment]
@ravisantoshgudimetla it'd be great to get this PR ready for review. Would you like help with that?

File: content/en/blog/_posts/2022-07-27-pod-os-field/index.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# Identify Pod Operating System authoritatively during pod admission time
+---
+layout: blog
+title: "Identifying pod operating system authoritatively during pod admission time"
+date: 2022-07-27
```
Comment (reylejano) [inline content/en/blog/_posts/2022-07-27-pod-os-field/index.md:5]:
```suggestion
date: 2022-09-05
```
Also please rename the file to 2022-09-05-pod-os-field.md

**Comment by reylejano** [comment]
Hi @ravisantoshgudimetla , this blog post is scheduled to be published next week on Sept 5.
Please have this PR at a "ready-state" (remove [WIP] from the title) and take a look at comments and suggestions

**Comment by sftim** [comment]
(if this is ready for review, please edit the PR title or add a comment to make it clear that this work should move forward)

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+The pod security standards `restricted` profile has been updated to ignore the following policies 
+- AllowPrivilegeEscalation
+- Capabilities
+- SeccompProfile
+when `.spec.os.name` for a Pod to `Windows`.
+
+
+## How do I use it?
+By setting `pod.spec.os.name=Windows` or `pod.spec.os.name=Linux`. Sample pod.yaml
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:46]:
```suggestion
By setting `spec.os.name: "Linux"` or `spec.os.name: "Windows"`. Here's a sample manifest:
```

(We list product from 3rd parties in alphabetical order, usually).

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+spec:
+  os:
+    name: Windows
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:55]:
```suggestion
    name: Windows # Make sure to use the correct case ("windows" is incorrect)
``

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    image: nginx:1.14.2
+    ports:
+    - containerPort: 80
+
+```
+
+## How can I learn more?
+KEP: https://github.com/kubernetes/enhancements/issues/2802
+API Changes: https://github.com/kubernetes/kubernetes/pull/104693
+Docs: https://kubernetes.io/docs/concepts/workloads/pods/#pod-os
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:67]:
```suggestion

- [KEP](https://github.com/kubernetes/enhancements/issues/2802)
- [k/k pull request #104693](https://github.com/kubernetes/kubernetes/pull/104693) (API changes for Pod OS)
- [Pod OS](https://kubernetes.io/docs/concepts/workloads/pods/#pod-os) in the Kubernetes documentation

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+## How do I get involved?
+This is a joint effort between [#sig-windows](https://kubernetes.slack.com/archives/C0SJ4AFB7), [#sig-auth](https://kubernetes.slack.com/archives/C0EN96KUY) and [#sig-node](https://kubernetes.slack.com/archives/C0BP8PW9G). 
+    Please reach out to us on those slack channels.
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:71]:
```suggestion
This is a joint effort between SIG Windows, SIG Auth, and SIG Node.  
Please reach out to us on those Slack channels:  [#sig-windows](https://kubernetes.slack.com/archives/C0SJ4AFB7), [#sig-auth](https://kubernetes.slack.com/archives/C0EN96KUY) and [#sig-node](https://kubernetes.slack.com/archives/C0BP8PW9G). 
```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+layout: blog
+title: "Identifying pod operating system authoritatively during pod admission time"
+date: 2022-09-05
+slug: pod-os-field-explained
+---
+
+**Authors:** Ravi Gudimetla (Apple)
+
+This article describes how to identify pod's operating system authoritatively using the `os` field in the pod spec, and how that
+helps in securing the kubernetes clusters while ensuring that the workloads can run smoothly.
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:11]:
```suggestion
helps in securing Kubernetes clusters while ensuring that the workloads can run smoothly.
```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+This article describes how to identify pod's operating system authoritatively using the `os` field in the pod spec, and how that
+helps in securing the kubernetes clusters while ensuring that the workloads can run smoothly.  
+
+## What problems does this solve?
+
+In the first releases of Kubernetes the `kube-apiserver` did not record any detail about which operating system a Pod should run on. Initially, Kubernetes only supported Linux
+nodes; later, Kubernetes added support for Windows nodes (stable since Kubernetes v1.14).
+You could use [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/)
+to give the scheduler a hint about where to place your Pod, but the actual Pod API didn't track that
+OS at all.
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:19]:
```suggestion
OS at all.

```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+As a result, the end user of a Kubernetes cluster can specify Linux specific security constraints onto Windows pods, or vice-versa.
+These security constraints are passed on to the underlying container runtime from kubelet and since these security constraints 
+are specific to OS, the containers may fail to come up.
+This problem worsens if you use admission plugins or webhooks like `PodSecurity` in your cluster,
+which validates the
+security constraints to the pods based on the security profile we choose.
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:25]:
```suggestion
security constraints to the pods based on the Pod security standard you choose to enforce.

```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+While the `kubelet` can identify the OS on which it runs and can strip certain security constraints before passing
+onto the container runtime, it'd be better to identify the OS early in the life-cycle of pod and use the 
+identification mechanism consistently across all kubernetes components including `kubelet`.
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:28]:
```suggestion
identification mechanism consistently across all Kubernetes components including `kubelet`. 
```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+These security constraints are passed on to the underlying container runtime from kubelet and since these security constraints 
+are specific to OS, the containers may fail to come up.
+This problem worsens if you use admission plugins or webhooks like `PodSecurity` in your cluster,
+which validates the
+security constraints to the pods based on the security profile we choose.
+While the `kubelet` can identify the OS on which it runs and can strip certain security constraints before passing
+onto the container runtime, it'd be better to identify the OS early in the life-cycle of pod and use the 
+identification mechanism consistently across all kubernetes components including `kubelet`. 
+With that in mind, `os` field has been added to the pod spec in 1.23 release of kubernetes as an 
+alpha feature and it graduated to stable in 1.25. In addition, the `PodSecurity` admission plugin has been updated to use the `os` field.
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:30]:
```suggestion
With that in mind, an `os` field was to the pod API in 1.23 release of Kubernetes (as an 
alpha feature). With Kubernetes v1.25, that field is stable, andthe `PodSecurity` admission
plugin has been updated to take the `os` field into account.
```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+onto the container runtime, it'd be better to identify the OS early in the life-cycle of pod and use the 
+identification mechanism consistently across all kubernetes components including `kubelet`. 
+With that in mind, `os` field has been added to the pod spec in 1.23 release of kubernetes as an 
+alpha feature and it graduated to stable in 1.25. In addition, the `PodSecurity` admission plugin has been updated to use the `os` field. 
+
+
+## How does it work?
+Pods now have an optional `.spec.os` field (and this is a stable feature, available in every cluster running
+Kubernetes v1.25 or later). Every object in Kubernetes is validated before it . Every object in kubernetes is validated before it 
+gets persisted to etcd during API admission time, including pod object.  Kubernetes v1.25 also introduced new validation which forbids Linux specific constraints to be set on Windows pods and vice-versa.
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:36]:
```suggestion
Kubernetes v1.25 or later). Every object in Kubernetes is validated before it . Every object in Kubernetes is validated before it 
gets persisted to etcd during API admission time, including Pods.  Kubernetes v1.25 also introduced new validation which forbids you from specify Linux-specific security controls for Windows pods, or vice-versa.
```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+The pod security standards `restricted` profile has been updated to ignore the following policies 
+- AllowPrivilegeEscalation
+- Capabilities
+- SeccompProfile
+when `.spec.os.name` for a Pod to `Windows`.
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:42]:
```suggestion
when you set the `.spec.os.name` for a Pod to `Windows`.
```

**Review by sftim — COMMENTED** [review]
Thanks. Some feedback that I recommend applying.

**Comment by sftim** [comment]
@katcosgrove let's postpone this one a week - does that work?

**Comment by katcosgrove** [comment]
@sftim We can do September 12!

**Comment by sftim** [comment]
OK, let's aim for publication on the 12th. @ravisantoshgudimetla would you be willing to work towards that?

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
+---
+layout: blog
+title: "Identifying pod operating system authoritatively during pod admission time"
```
Comment (sftim) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:3]:
```suggestion
title: "Identifying Pod Operating System Authoritatively During Pod Admission Time"
```

**Review by sftim — COMMENTED** [review]
We'll need to pick a new publication date. How about the 20th of September?

@ravisantoshgudimetla you will need to edit this PR to use that date. If that's not feasible, an editor (possibly me) may select a date and push commits to this PR so that we can publish it, separate from the Kubernetes v1.25 release blog series.

**Comment by ravisantoshgudimetla** [comment]
> @ravisantoshgudimetla you will need to edit this PR to use that date. If that's not feasible, an editor (possibly me) may select a date and push commits to this PR so that we can publish it, separate from the Kubernetes v1.25 release blog series.

Thanks for the help @sftim. Updated the publication date.

**Comment by sftim** [comment]
@ravisantoshgudimetla we're happy to have a version of this article for the blog, but it has missed the timings for post-release blogs. Even though the timing didn't work out, I'd like to make use of the work that has gone in.

Let's aim for a new date. How about the 19th of October? Does that work.

You need to make two changes: one to the filename, and another to the date in the front matter.
Could you also recommend someone who can review the content for technical accuracy? We need to review this before we publish it, and that review needs to cover the technical side.

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
+---
+layout: blog
+title: "Identifying Pod Operating System Authoritatively During Pod Admission Time"
+date: 2022-09-20
```
Comment (reylejano) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:4]: The published date has passed, the filename also needs to be changed
```suggestion
date: 2022-10-11
```

**Comment by reylejano** [comment]
Hi @kubernetes/sig-windows-leads, can we get a technical review for this blog
Thank you

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+**Authors:** Ravi Gudimetla (Apple)
+
+This article describes how to identify pod's operating system authoritatively using the `os` field in the pod spec, and how that
+helps in securing Kubernetes clusters while ensuring that the workloads can run smoothly.
+
+## What problems does this solve?
+
+In the first releases of Kubernetes the `kube-apiserver` did not record any detail about which operating system a Pod should run on. Initially, Kubernetes only supported Linux
+nodes; later, Kubernetes added support for Windows nodes (stable since Kubernetes v1.14).
+You could use [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/)
```
Comment (marosset) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:17]: Maybe we should just say there are many different ways to give the scheduler a hint?
One of the more common ways is to use a nodeSelector.

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+As a result, the end user of a Kubernetes cluster can specify Linux specific security constraints onto Windows pods, or vice-versa.
+These security constraints are passed on to the underlying container runtime from kubelet and since these security constraints 
+are specific to OS, the containers may fail to come up.
+This problem worsens if you use admission plugins or webhooks like `PodSecurity` in your cluster,
+which validates the
+security constraints to the pods based on the Pod security standard you choose to enforce.
+
+While the `kubelet` can identify the OS on which it runs and can strip certain security constraints before passing
+onto the container runtime, it'd be better to identify the OS early in the life-cycle of pod and use the
```
Comment (marosset) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:29]:
```suggestion
onto the container runtime, it'd be better to identify the target OS early in the life-cycle of pod and use the 
```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+are specific to OS, the containers may fail to come up.
+This problem worsens if you use admission plugins or webhooks like `PodSecurity` in your cluster,
+which validates the
+security constraints to the pods based on the Pod security standard you choose to enforce.
+
+While the `kubelet` can identify the OS on which it runs and can strip certain security constraints before passing
+onto the container runtime, it'd be better to identify the OS early in the life-cycle of pod and use the 
+identification mechanism consistently across all Kubernetes components including `kubelet`. 
+With that in mind, an `os` field was to the pod API in 1.23 release of Kubernetes (as an 
+alpha feature). With Kubernetes v1.25, that field is stable, andthe `PodSecurity` admission
```
Comment (marosset) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:32]:
```suggestion
alpha feature). With Kubernetes v1.25, that field is stable, and the `PodSecurity` admission
```

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+plugin has been updated to take the `os` field into account.
+
+
+## How does it work?
+Pods now have an optional `.spec.os` field (and this is a stable feature, available in every cluster running
+Kubernetes v1.25 or later). Every object in Kubernetes is validated before it . Every object in Kubernetes is validated before it
```
Comment (marosset) [inline content/en/blog/_posts/2022-09-05-pod-os-field/index.md:38]:
```suggestion
Kubernetes v1.25 or later). Every object in Kubernetes is validated before it 
```

**Comment by marosset** [comment]
One of the big motivating factors of having this pod os field is so that windows pods can still be scheduled with `Restricted` policies which require things like "containers must drop ALL capabilities". 
This is called out https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted

I think it would be good to highlight this in the blog post.

**Comment by sftim** [comment]
@ravisantoshgudimetla would you be willing to to revise this in light of https://github.com/kubernetes/website/pull/35439#issuecomment-1267667477 ?

**Comment by tengqm** [comment]
@ravisantoshgudimetla Ready to give this another stab?

**Comment by NitishKumar06** [comment]
@ravisantoshgudimetla Ping!

**Comment by claudiubelu** [comment]
Ping @ravisantoshgudimetla . If you're not available, we can take over this PR and address the necessary changes.

File: content/en/blog/_posts/2022-09-05-pod-os-field/index.md (hunk 18 above)
Comment (ravisantoshgudimetla) [reply content/en/blog/_posts/2022-09-05-pod-os-field/index.md:17]: I'll reword it to `While Kubernetes has multiple ways to hint scheduler about the node on which pod needs to run, a more common way is to use nodeSelector`, WDYT?

**Comment by ravisantoshgudimetla** [comment]
> One of the big motivating factors of having this pod os field is so that windows pods can still be scheduled with Restricted policies which require things like "containers must drop ALL capabilities".
This is called out https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted
> I think it would be good to highlight this in the blog post.

I have added it here - https://github.com/kubernetes/website/pull/35439/files#diff-f2f020ab42d6a660ddb4835f70a43f8a64f9f50204e7ab49d0f275d76af4345cR31. PTAL

**Comment by sftim** [comment]
@ravisantoshgudimetla how about a new publication date - perhaps the 14th of March 2023? If you would like to aim for that, we can see if we can get reviews to happen on the blog team side.

**Comment by marosset** [comment]
@ravisantoshgudimetla - can you make me a collaborator on your k/webite fork and then I can help push this through?

**Comment by divya-mohan0209** [comment]
@ravisantoshgudimetla : Please advise if you'd like to continue work on this PR and also, let us know of a realistic target date for publication. Additionally, it'd be great if you could grant collaborator access to @marosset as requested in the above comments. This has been in the works for a year and I'm sure you'd appreciate that we'd like to see this merged sooner rather than later.

**Comment by divya-mohan0209** [comment]
Hey @ravisantoshgudimetla , thank you for all your hard work on this! However, since there has been no activity on addressing the feedback despite repeated reminders we'll be closing the PR. Please feel free to reopen the PR whenever you have the bandwidth to work together so that we can get this blog published. 
/close
